### ISSUE TITLE
Native Library Loader API

### ISSUE THREAD

**Issue by annaaniol** [dotnet/corefx#32015]
This issue introduces a NativeLibrary abstraction for loading native libraries from managed code in a platform independent way. It provides two key functionalities

- A facility to register a callback to implement custom native library load logic from managed codeper assembly. 
- A few methods to handle native library manipulations from managed code.

### Context
This API is introduced as part of implementing the [Dllmap](https://github.com/dotnet/coreclr/pull/19373) feature. This issue introduces a generic callback strategy, where DllMap is one specific implementation using the API.

### Approved API

```C#
namespace System.Runtime.InteropServices
{
   /// <summary>
   /// A delegate used to resolve native libraries via callback.
   /// </summary>
   /// <param name="libraryName">The native library to resolve</param>
   /// <param name="assembly">The assembly requesting the resolution</param>
   /// <param name="DllImportSearchPath?">The DllImportSearchPathsAttribute on the PInvoke, if any. 
   ///         Otherwise, the DllImportSearchPathsAttribute on the assembly, if any. Otherwise null.
   /// </param>
  /// <returns>The handle for the loaded native library on success, null on failure</returns>  
   public delegate IntPtr DllImportResolver(string libraryName,
                                             Assembly assembly,
                                             DllImportSearchPath? searchPath);

    public static partial class NativeLibrary
    {
        // Typical or for dependencies which must be there or failure should happen.

        public static IntPtr Load(string libraryPath);
        public static IntPtr Load(string libraryName,
                                  Assembly assembly,
                                  DllImportSearchPath? searchPath);

        // For fast probing scenarios:

        public static bool TryLoad(string libraryPath,
                                   out IntPtr handle);
        public static bool TryLoad(string libraryName,
                                   Assembly assembly,
                                   DllImportSearchPath? searchPath,
                                   out IntPtr handle);

        public static void Free(IntPtr handle);

        public static IntPtr GetExport(IntPtr handle, string name);
        public static bool TryGetExport(IntPtr handle, string name, out IntPtr address);

       /// <summary>
       /// Set a callback for resolving native library imports from an assembly.
       /// This per-assembly callback is the first attempt to resolve native library loads 
       /// initiated by this assembly.
       ///
       /// Only one resolver callback can be registered per assembly. Trying to register a second
       /// callback fails with InvalidOperationException.
       /// </summary>
       /// <param name="assembly">The assembly for which the callback is registered</param>
       /// <param name="resolver">The callback to register</param>
       /// <exception cref="System.ArgumentNullException">If assembly or callback is null</exception>
       /// <exception cref="System.InvalidOperationException">If a callback is already set for this assembly</exception>
        public static void SetDllImportResolver(Assembly assembly, DllImportResolver resolver);
    }
}
```

### Related

- [Helper class for dealing with native shared libraries and function pointers](
 https://github.com/dotnet/corefx/issues/17135)
- [Add NativeLibrary class](https://github.com/dotnet/coreclr/pull/16409)
- [DllMap Design](https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/dllmap.md)

### Earlier Proposal
```C#
namespace System.Runtime.InteropServices
{
    /// <summary>
    ///  NativeLibrary class is an abstraction for loading native libraries
    ///  from managed code in a platform independent way.
    /// </summary>
    public class NativeLibrary
    {
        /// <summary>
        ///  NativeLibrary constructor, given a name and a library handle
        /// </summary>
        /// <param name="libraryName">The library Name</param>
        /// <param name="handle">The library handle obtained from the appropriate system call</param>
        /// <exception cref="System.ArgumentNullException">Thrown when libraryName equals null or handle equals IntPtr.Zero.</exception>
        public NativeLibrary(string libraryName, IntPtr handle)
        {
            Name = libraryName ?? throw new ArgumentNullException("Name of a NativeLibrary can't be set to null.");
            if (handle == IntPtr.Zero)
                throw new ArgumentNullException("Handle of a NativeLibrary can't be set to IntPtr.Zero.");
            Handle = handle;
        }

        /// <summary>
        /// The simple Library loader 
        /// This is a wrapper around OS loader, using "default" flags
        /// This method facilitates platform independent native library load in the simple case
        /// More complex loads will need to PInvoke to the platform-specific system calls.
        /// </summary>
        /// <param name="libraryName">The name of the native library to be loaded</param>
        /// <returns>The NativeLibrary object for the loaded library</returns>  
        /// <exception cref="System.DllNotFoundException ">Thrown if the library can't be found.</exception>
        /// <exception cref="System.BadImageFormatException">Thrown if the library is not valid.</exception>
        public static NativeLibrary Load(string libraryName);

        /// <summary>
        /// Load a native library
        /// This loader API exposes high-level functionality. Given a library name, it searches 
        /// specific paths based on runtime configuration and attributes of the calling module.
        /// This Load() method works at a 'lower-level' than the managed call-backs for native 
        /// library loading. That is, the NativeLibrary.Load() methods do not call back into
        /// * The Per-assembly function registered via NativeLibrary.RegisterNativeLibraryLoadCallback()
        /// * Or AssemblyLoadContext.LoadUnmanagedDll()
        /// </summary>
        /// <param name="libraryName">The name of the native library to be loaded</param>
        /// <param name="dllImportSearchPath">The search path</param>
        /// <param name="assembly">The assembly loading the native library</param>
        /// <returns>The NativeLibrary object for the loaded library</returns>  
        /// <exception cref="System.DllNotFoundException ">Thrown if the library can't be found.</exception>
        /// <exception cref="System.BadImageFormatException">Thrown if the library is not valid.</exception>
        public static NativeLibrary Load(string libraryName, DllImportSearchPath dllImportSearchPath, Assembly assembly);

        /// <summary>
        /// Close a Native library
        /// This function is a simple wrapper around platform-specific OS calls.
        /// </summary>
        /// <returns>True, if the operation succeeds</returns>  
        /// <exception cref="System.InvalidOperationException">If the NativeLibrary object is not backed by an open handle</exception>
        public bool Free();

        /// <summary>
        /// Get the address of a Symbol
        /// This is a simple wrapper around OS calls, and does not perform any name mangling.
        /// </summary>
        /// <returns>The address of the symbol, if it exists in this NativeLibrary</returns>  
        /// <exception cref="System.InvalidOperationException">If the NativeLibrary object is not backed by an open handle</exception>
        /// <exception cref="System.ArgumentException">If the symbol is not found</exception>
        public IntPtr GetProcAddress(string name);

        /// <summary>
        ///  The actual library handle obtained via LoadLibrary() or dlOpen() operations
        /// </summary>
        public IntPtr Handle { get; }

        /// <summary>
        /// The name referenced by the importing assembly
        /// ToString() can return more detailed/debugging information (ex: mapped-name, loaded location, etc)
        /// </summary>
        public string Name { get; } // Will Return the name referenced by the importing assembly, set only for debugging purpose.

        /// <summary>
        /// Method to register a callback to perform NativeLibraryLoads
        /// The callbacks are registered per-assembly, and can implement custom native library load logic
        /// At most one callback can be registered per assembly
        /// Callbacks can be unregistered by passing a null 'callback' parameter
        /// </summary>
        /// <param name="assembly">Assembly with which this callback should be associated </param>
        /// <param name="callback">The callback function</param>
        /// <returns>True, if the callback was updated</returns>  
        /// <exception cref="System.ArgumentNullException">If the assembly argument is null</exception>
        /// <exception cref="System.InvalidOperationException">Thrown when there is already a callback registered for the specified assembly.</exception>
        public static bool RegisterNativeLibraryLoadCallback(Assembly assembly, Func<string, DllImportSearchPath, Assembly, NativeLibrary> callback);
    }
}
```

**Comment by AaronRobinsonMSFT** [comment]
cc: @jeffschwMSFT @luqunl @swaroop-sridhar

**Comment by swaroop-sridhar** [comment]
I think the title shouldn't say "Dllmap specific events." 
While the API is being proposed in the context of implementing DllMap, the APIs are designed to be generically useful for anyone who wants to customize loading of native binaries -- and should be reviewed in that context. 

For example, the `RegisterGetEntrypointCallBack` may be used to locate entry-points that are embedded/linked within a bundled executable. (cc: @sbomer @sdmaclea)

**Comment by swaroop-sridhar** [comment]
I also think that GetEntrypointDelegate() should take a NativeLibrary handle, since this information is unique enough to identify all other information, if necessary.

**Comment by swaroop-sridhar** [comment]
We need two pieces of information from `NativeLibrary` for diagnostic purposes. 
- The  `Name()` or `ToString()` which refers to the name by which the calling-assembly refers to the imported library
- the `Path()`  to the actual native library that got loaded (say, after DllMap).

**Comment by Wraith2** [comment]
Can a native library be unloaded? 
Does the dll/so/dylib have to be located on storage or can they be memory resident only, in which case path wouldn't be distinguishing feature.
Even though the `Name` is for diagnostic purposes it may be wanted at runtime for logging etc, the debugger is not the only way to want to inspect this so I support keeping it.
Events vs callbacks doesn't really matter as long as both provide multiple subscribers, `AppDomain.AssemblyResolve` is an event for historic reasons so I would tend to side with the event mechanism for consistency but any technical merit could outweigh that.

**Comment by svick** [comment]
* What is the relationship between this proposal and https://github.com/dotnet/corefx/issues/17135, which is marked as approved?
* For the events version, is it okay that it doesn't follow [Framework Design Guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/event) by not using `EventHandler<T>`, `EventArgs` and having non-`void` return type?
* For the callbacks version, why do the `Register` methods return the delegate back? And is it okay that there doesn't seem to be any way to unregister?

**Comment by jkotas** [comment]
> the Path() to the actual native library that got loaded 

You cannot always tell the path that got actually loaded.

**Comment by AaronRobinsonMSFT** [comment]
>What is the relationship between this proposal and #17135, which is marked as approved?

@svick, That is a good point. We need to reconcile these two proposals. cc @danmosemsft and @mellinoe.

**Comment by AaronRobinsonMSFT** [comment]
>Can a native library be unloaded?

I am inclined to say 'no'. The CLR has a number of caching mechanisms for native library scenarios and I don't know how easy it would be to properly update all of them. There is the native library cache that is for load library and we would also need to look at `DllImport` and make sure all related native method descs can be removed. There could be more in the CLR and that is not even getting to the scenario where users could cache delegates. I don't see an easy path forward to support unloading from the API itself.

**Comment by jeffschwMSFT** [comment]
@jkotas 
> Should GetEntrypointDelegate take the library handle as an argument; not library name?

The intention of providing the non-modified version of libraryname and entrypoint from the dllmap is to provide the right information to the developer to choose to either return the symbol as declared in code or find a suitable replacement.  The NativeLibrary would then be completely managed by the developer and the runtime/fx would not use it as an exchange type.

> For the callback alternative, the existing Marshal or NativeLibrary would be a good place for it and it can be a single method for both

We discussed the callback mechanism with the caveat that it would likely use an override semantic and only respect the last callback provided for a given assembly.  In that case returning a potenial overriden callback seemed important.

@Wraith2 
> Events vs callbacks doesn't really matter as long as both provide multiple subscribers

The current design is ambigous on this point, but the initial thought would be that events are multi subscriber but callbacks would be single with the caller having to chain if one is overwritten.

@svick, @AaronRobinsonMSFT 
>> What is the relationship between this proposal and #17135, which is marked as approved?
> @svick, That is a good point. We need to reconcile these two proposals. cc @danmosemsft and @mellinoe.

This is the reconcilation, eg. new version.  We are drafting off of the approved proposal but there are differences in how complete a solution we are proposing.

**Comment by annaaniol** [comment]
@svick I updated the events version to follow the [Framework Design Guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/event):

``` c#
namespace System.Reflection
{
    public abstract partial class Assembly
    {
        public static event EventHandler<LoadNativeLibraryArgs> LoadNativeLibrary;
        public static event EventHandler<GetEntrypointArgs> GetEntrypoint;
    }

    public class LoadNativeLibraryArgs : EventArgs
    {
        public LoadNativeLibraryArgs(string libraryName, DllImportSearchPath dllImportSearchPath, Assembly callingAssembly)
        {
            LoadedAssembly = loadedAssembly;
            DllImportSearchPath = dllImportSearchPath;
            CallingAssembly = callingAssembly;
        }

        public string LibraryName { get; set; }
        public DllImportSearchPath DllImportSearchPath { get; set; }
        public Assembly CallingAssembly { get; set; }
    }

    public class GetEntrypointArgs : EventArgs
    {
        public GetEntrypointArgs(string libraryName, string entrypointName)
        {
            LoadedAssembly = loadedAssembly;
            EntrypointName = entrypointName;
        }

        public string LibraryName { get; set; }
        public string EntrypointName { get; set; }
    }
}


namespace System
{
    public delegate void LoadNativeLibraryEventHandler(object sender, LoadNativeLibraryArgs args);
    public delegate void GetEntrypointEventHandler(object sender, GetEntrypointArgs args);
}
```

**Comment by annaaniol** [comment]
@jkotas 
> Do we know how often is the entrypoint remapping done via dllmap today? It may be useful to collect some statistics about it.

How can we collect such statistics?
I'd say such a simplification will strongly limit the functionality which is we don't want to happen.

**Comment by jkotas** [comment]
> How can we collect such statistics?

@jeffschwMSFT should be able to help to connect you with the right people. We have data mining experts who should be able to do this easily.

**Comment by jeffschwMSFT** [comment]
@jkotas 
> Do we know how often is the entrypoint remapping done via dllmap today?

I feel that the entrypoint portion is pretty important.  Regarding the name mangling, if GetNativeEntryPoint provides the same level of name mangling as the original p/invoke, won't that be sufficient?

> @jeffschwMSFT should be able to help to connect you with the right people. We have data mining experts who should be able to do this easily.

A quick search on github for dllmap and dllentry yields hits for both.  I agree with Jan that many of them are just for a dll, but there are certainly plenty (that is a scientific measurement) that have name and target.  

https://github.com/search?p=2&q=dllentry&type=Code

**Comment by masonwheeler** [comment]
@Wraith2 Yes?

There doesn't seem to be much of anything here that impacts my reasoning wrt why DllMap is a bad idea that should not be incorporated into .Net Core.

**Comment by swaroop-sridhar** [comment]
>> You cannot always tell the path that got actually loaded.
Path is probably a bad name (for assets like memory mapped loads etc) . However, we'll need some string to report the information here (mapped name / path / memory identifier, etc.) for diagnostic purpouses

**Comment by swaroop-sridhar** [comment]
@jkotas had recommended using Func<> notation instead of delegate typenames. So, the callback-registration version should be something like:
```
namespace System.Runtime.InteropServices
{
    public sealed partial class NativeLibrary
    {
          ...
         public static Func<string, DllImportSearchPath, Assembly> RegisterNativeLibraryLoadCallBack(Func<string, DllImportSearchPath, Assembly> callBack);
          ...
    }
}
```

**Comment by jkotas** [comment]
> if GetNativeEntryPoint provides the same level of name mangling as the original p/invoke, won't that be sufficient?

I agree that we can make the entrypoint remapping work if we need to. My question is whether it is useful enough to add it to .NET Core. So far, I am failing to find a set of good examples of what it is actually needed for. It feels like a corner case that it is not worth to have and the few libraries that use it can workaround.

**Comment by jeffschwMSFT** [comment]
Perhaps I am missing the cost of getting entry point mapping working.  We could expose the equivalence of https://github.com/dotnet/coreclr/blob/master/src/vm/dllimport.cpp#L5683 as GetEntryPoint.

Another thought on why entry point may be important is for the DSL case that Miguel pointed out.  If you wanted to put your own encoding of library and entrypoint in the dllimport you would need the pair to complete that scenario.

My only concern is that dllmap with only libraries may not be complete enough.

**Comment by annaaniol** [comment]
So we agreed on the following things:
- a native library can't be unloaded because of the reasons that @AaronRobinsonMSFT has pointed out,
- we will use `LoadNativeLibrary` and `GetEntrypoint` events (using EventHandler<T>),
- we keep exposing `GetNativeEntrypoint`,
- `toString()` on `NativeLibrary` will return debugging string that includes source library name, target name that was loaded, and location details.

I updated the description.

**Comment by annaaniol** [comment]
I included your suggestions and updated the Proposed API. Please take a final look and tell me if you have additional comments or questions.

**Comment by jkotas** [comment]
These comments do not appear to be addressed:

- The methods to register the callback should not be on Assembly itself. We do not want to add interop-specific methods to Assembly. https://github.com/dotnet/corefx/issues/32015#issuecomment-416819593
- This is not providing complete information to resolve the entrypoint. It would also need to provide the signature of the method and the method attributes to do the right name mangling. https://github.com/dotnet/corefx/issues/32015#issuecomment-417054060 Alternative would be to drop the entrypoint remapping from the proposal. I am not convinced based on the dllmap usage data that we need it - I know that we do not have agreement on this yet.

**Comment by jeffschwMSFT** [comment]
@jkotas 
> The methods to register the callback should not be on Assembly itself. We do not want to add interop-specific methods to Assembly

Anna and I discussed this.  I think the challenge is that to achieve the proper Assembly level isolation we need to associate the callback with an assembly.  The simplest approach would be too extend Assembly, but I hear your feedback.  

I think we should align with your recommendation, which moves us away from events and back to callbacks.  Within that design we felt that we would only allow one callback and then return any potentially overwritten callback.  Similar to @swaroop-sridhar suggestion https://github.com/dotnet/corefx/issues/32015#issuecomment-417106670 (two seperate callbacks, using the Func notation and returns a potentially overwritten callback, NativeLibrary would then retain a mapping of Assembly with the callbacks).

> This is not providing complete information to resolve the entrypoint. It would also need to provide the signature of the method and the method attributes to do the right name mangling

I thought that based on our discussion that GetEntryPoint would be provided for the simple case.  The runtime would then use the method and method attributes provided in the original dllimport to mangle as necessary.  In the event that is not sufficient the developer could provide a pre-managed name to the API (we will ensure that pattern works with our implementation).  In essese we will expose a p/invoke to GetProcAddress/GetSymbol in a way that does not introduce a cycle with the dllmap hook.

**Comment by swaroop-sridhar** [comment]
I did some more search for Dll entry point map usage (outside of github), and I did not see widespread use of them. Most of the projects seem to use it for Mono specific mapings. In particular: 
```
<dllmap dll="i:kernel32.dll">
  <dllentry dll="__Internal" name="CopyMemory" target="mono_win32_compat_CopyMemory"/>
  <dllentry dll="__Internal" name="FillMemory" target="mono_win32_compat_FillMemory"/>
  <dllentry dll="__Internal" name="MoveMemory" target="mono_win32_compat_MoveMemory"/>
  <dllentry dll="__Internal" name="ZeroMemory" target="mono_win32_compat_ZeroMemory"/>
</dllmap>
```
I tend to agree with Jan that the reason these mappings are not widely used is that they only work in the situation where we have Dlls on different platforms that differ in name but not in signature. 

The place where I can imagine this being useful is: A developer implements the same native code with different naming conventions on different platforms. For example, it is common to name public functions in the style `DoSomething()` on Windows and `doSomething()` on Unix – which may be enforced by stylecop like analyzers. But we haven’t seen this usage in publicly searchable code.

If we want to support entrypoint maps, we'll need to support remangling the name. The method-descriptor has pre-computed the mangled name wrt the original name, and we need to compute it based on the mapped name in order to execute `GetProcAddress()`

We’ll need to do one of:
•	Pass signature-info (likely as a method-desc reflection object) object to the GetEntryPoint event handler. 
•	Get only the name of the new entryPoint from the handler, and compute the mangled name in the native code
•	Expose a service to compute the mangled name from managed code

Of these, I think the first option is more generic -- ex: Some other implementation of GetEntryPoint handler may make decisions based on the signature.

**Comment by jeffschwMSFT** [comment]
> Get only the name of the new entryPoint from the handler, and compute the mangled name in the native code

Do we feel this would not be a viable option?  The existing NDirectEntryPoint uses the method descriptor in the p/invoke to compute these values.  If the API were to provide a different set of method information would that even work?

**Comment by swaroop-sridhar** [comment]
>> Do we feel this would not be a viable option? 

Yes, I think we can get this to work. Since the native code used some utility to compute the first mangled name, we should be able to compute the new name too. We talked about this with @jkotas yesterday.

**Comment by jkotas** [comment]
> Yes, I think we can get this to work. Since the native code used some utility to compute the first mangled name, we should be able to compute the new name too. 

You can make it work, but the design for it is not obvious. My recomendation would be to drop the entrypoint remapping callback from this phase, and just have callback for the native library loading. We can add the entrypoint callback later if it is proven to be really required.

**Comment by AaronRobinsonMSFT** [comment]
@jeffschwMSFT  I am inclined to agree with @jkotas about the name remapping if only because of the intended scope of @annaaniol 's project. It should be relatively easy to enrich the API if the community demands this functionality prior to the .NET Core 3.0 release.

**Comment by swaroop-sridhar** [comment]
Another option to consider is whether to have get/set APIs. That is, instead of 
```C#
public static Func<string, DllImportSearchPath, Assembly> RegisterNativeLibraryLoadCallback(Assembly assembly, Func<string, DllImportSearchPath, Assembly> callback);
```
consider
```C#
public static Func<string, DllImportSearchPath, Assembly> GetNativeLibraryLoadCallback(Assembly assembly);
public static bool SetNativeLibraryLoadCallback(Assembly assembly, Func<string, DllImportSearchPath, Assembly> callback);
```
The later version will facilitate writing code like:
```C#
if (GetNativeLibraryLoadCallback(assembly) == null)
{
    SetNativeLibraryLoadCallback(assembly, callback);
}
```
for chaining callbacks
```C#
oldCallback = GetNativeLibraryLoadCallback(assembly);

SetNativeLibraryLoadCallback(assembly, (oldCallback == null) ? newCallback : ( (d,p,a) => { oldCallback(d,p,a); newCallback(d,p,a);}  );
```
or for synchronization
```C#
if (GetNativeLibraryLoadCallback(assembly) == null)
{
    Lock();
    if (GetNativeLibraryLoadCallback(assembly) == null)
    {
        SetNativeLibraryLoadCallback(assembly, callback);
    }
    Unlock();
}
```
However, this issue is not critical, because I think having multiple callbacks for the same assembly is rare, and that the behavior similar to `get/setCallBack` can be achieved using the `RegisterCallback` version.

**Comment by annaaniol** [comment]
I updated the description by removing `NativeLibrary::GetEntrypoint()` and the event-based vision. I think we agreed on having callbacks and exposing only the `NativeLibrary::Load()` function & mapping only libraries when using the dllmap. I also added the `assembly` argument to `RegisterNativeLibraryLoadCallback` as Swaroop suggested.

**Comment by jkotas** [comment]
> `Func<string, DllImportSearchPath, Assembly>`

This seems to be missing return type. What is the callback going to return?

> `static Func<string, DllImportSearchPath, Assembly> RegisterNativeLibraryLoadCallback`

It is unusual API design for Register method to return the previous callback. I do not think there is any prior art like that - here are examples of the existing Register methods: https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken.register?view=netcore-2.1 and https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.registerprovider?redirectedfrom=MSDN&view=netcore-2.1#System_Text_Encoding_RegisterProvider_System_Text_EncodingProvider_

I understand that you are doing this to allow more than one callback for given assembly. Do we really need to allow more than one callback for given assembly? If we really do, why not let the runtime call each of them and stop once one of them returns non-null library handle?

**Comment by AaronRobinsonMSFT** [comment]
>It is unusual API design for Register method to return the previous callback.

@jkotas I agree this is not common the .NET world, but it is common in C++. See as an example: https://en.cppreference.com/w/cpp/error/set_terminate. Unsure if that sways the design in either direction, but I have always appreciated the ability to reset the value, but it may not matter in this case.

**Comment by jeffschwMSFT** [comment]
https://github.com/Microsoft/perfview/blob/9d8a9b90c5ed917da1b9bdea98210114a270e633/src/TraceEvent/TraceEvent.cs#L3212

This is an example of how TraceEvent handles a similar event where their could be multiple subscribers to a callback style event.  They are chained in a simple manner.

**Comment by jkotas** [comment]
Right, there are multiple choices for how to deal with multiple callbacks. Looking at real scenarios when we expect multiple callbacks to be registered should help us choose the right one.

**Comment by masonwheeler** [comment]
Wait... isn't this literally the entire point of `event` and `MulticastDelegate`?  What am I missing here?

**Comment by jeffschwMSFT** [comment]
There is a semantic difference, for event all handlers will be invoked.  Here once we get a result we are going to stop calling the callbacks.

**Comment by annaaniol** [comment]
I updated the description. There is a corresponding PR with the implementation: [https://github.com/dotnet/coreclr/pull/19826](https://github.com/dotnet/coreclr/pull/19826)

> Do we really need to allow more than one callback for given assembly? 

No, and in the current version, if `RegisterNativeLibraryLoadCallback` gets called multiple times for one assembly, the registered callback will be overridden by the last specified callback.

**Comment by AaronRobinsonMSFT** [comment]
> ` public string Name { get; set; }`

Should be `private set`. If at all, I thought the `Name` property was being removed, if not please mark the setter private.

**Comment by annaaniol** [comment]
> public string Name { get; private set; }

changed

**Comment by luqunl** [comment]
> public class CallbackAlreadyRegistered: System.Exception
Nit: Could you add Exception to this class name?

>  Name = libraryName ?? throw new ArgumentException("Name of a NativeLibrary can't be set to null.");
How about throw new ArgumentNullException?

**Comment by jkotas** [comment]
It is not worth it to add a new `CallbackAlreadyRegisteredException` type for this. This exception should be deleted from the proposal, and the API should throw InvalidOperationException with appropriate error message instead.

**Comment by annaaniol** [comment]
CallbackAlreadyRegisteredException has been removed and I replaced ArgumentException with ArgumentNullException.

**Comment by xoofx** [comment]
Hey, while reading this issue and the associated design documents, I have a few questions:

- Why a managed class `NativeLibrary` is being introduced for this? (e.g instead of having a low level API `IntPtr NativeLibrary.Load(...)` and `Free` as they match to the underlying OS API) 
- Is this class going to track the native dll and unload it when the `NativeLibrary` is no longer referenced? I assume that going with `static NativeLibrary Load(...)` would return the same instance? 
- When passing an assembly to it, it says _"The assembly loading the native library"_: What does it mean in practice? Is the assembly owning the NativeLibrary object, if multiple assemblies are loading the same native library, will we get multiple NativeLibrary instance? I suspect it is to tight it to an underlying ALC, but why not reference the ALC directly then?
- I understand that it is only for loading a dll, but what's going to be the API for resolving the function from it? Instance method on NativeLibrary likely? returning IntPtr and/or delegate?

I would have preferred to have lower level API first (easily xplat) and I would not mind having a higher level API (like NativeLibrary) on top of it, but going straight to the high level only feels wrong... I understand the requirement when the underlying OS APIs have really different behavior, but afaik, in that case, they are 1-1

**Comment by AaronRobinsonMSFT** [comment]
>I would have preferred to have lower level API first

How much lower would you like it? Can you elaborate on your definition of 'lower level'? This literally has the `Load(...)` function on it. The `NativeLibrary` class also exposes the associated `IntPtr` via the `Handle` property. The remaining functions are properties are for diagnostic and convenience purposes. The `Load(...)` also has the option to tag flags which will allow us to implement this in both Windows and non-Windows environments.

The `Free` option is definitely up for on-going debate. There are serious issues with providing that kind of functionality and it is unclear if it is a hard requirement or if it would just be a nice to have. There have been comments regarding load a version of a library and then unload it and load another version, but there hasn't been any hard numbers to confirm the need or what the actual gain would be other than some generic anecdotes. If you have scenarios with concrete numbers on the win and need for the functionality please provide them so the conversation can be pushed in one direction or another.

**Comment by xoofx** [comment]
> How much lower would you like it? Can you elaborate on your definition of 'lower level'? 

As described [here](https://github.com/dotnet/corefx/issues/17135#issuecomment-379444280)

```c#
IntPtr LoadNativeLibrary(string libraryPath);
IntPtr GetNativeSymbolAddress(IntPtr nativeLibraryHandle, string symbolName);
void CloseNativeLibrary(IntPtr nativeLibraryHandle);
```

These functions are typically mapping directly to direct OS calls on the OS I'm familiar with (e.g even on PS4) No higher level stuffs like `DllImportSearchPath`... etc.

Dllmap, dll search map...etc. are higher level for me. I'm really looking forward a dllmap solution, but I would like that things are decoupled and we don't conflate all problems into a single fat solution.

**Comment by xoofx** [comment]
> @xoofx What's the difference between an API that low and p/invoke?

I didn't say that it could not be p/invoke actually 😉 Usually, you need a very small indirection before going to p/invoke directly (because dlopen takes a flags argument for example...etc.), but yeah, it's nothing more than this. Just that you need a DllImport per platform to support it transparently (if you roll your own), so you end up today having different assemblies per platform or you need to if/else at runtime (which is not great, because detecting a platform from .NET is not always that easy)

**Comment by jkotas** [comment]
>  there hasn't been any hard numbers to confirm the need or what the actual gain would be other than some generic anecdotes. If you have scenarios with concrete numbers on the win and need for the functionality 

Search github for [FreeLibrary](https://github.com/search?l=C%23&q=FreeLibrary&type=Code) or [dlclose](https://github.com/search?l=C%23&q=dlclose&type=Code) use in C#. Some of the uses you will find are appropriate, some of them are questionable - that's what happens when there is ambiguous problem with no clear guidance.

Here are a few good examples:

https://github.com/2gis/nuclear-river-validation-rules/blob/003132cbf672ea6eac4d1e8257f00172ec7a841f/ValidationRules.Querying.Host/App_Start/LibRdKafkaLoader.cs#L47
https://github.com/slippyex/OSME/blob/c107f2b67a67be0ac2bb1525c1f579ba0a6ad473/OSMEngine.v.0.92/samples/VC%23/sample/sample/Program.cs#L13
https://github.com/robotdotnet/NetworkTablesCore/blob/d2c2f314df3095d166e802f7fb55fba4a41b6831/NetworkTablesCore/Native/Interop.cs#L127
https://github.com/esteve/ros2_dotnet/blob/e0ea001d2a63a6d23d440aed26f9223dc5459d2a/rcldotnet_common/DllLoadUtils.cs#L123

**Comment by AaronRobinsonMSFT** [comment]
@jkotas I have seen examples - the ones provided are also basically the same. I don't there are developers unloading native DLLs, more of my statement is around 'why?' - what scenario is being unblocked here, other than 'I want this'. The hard numbers I am looking for are performance, memory, some real numbers that move the needle on why this is important.

@xoofx We are not exposing the raw `LoadLibrary` or `dlopen` since it is so simple to just do a P/Invoke yourself. What we are providing from the `Load()` functions is how the CLR does library loading - with all the magic probing that comes along with it. I don't really see any real difference between the low level you are suggesting and the low level we have provided. Are you considering a scenario where you _don't_ want to let the CLR apply any magic look up help?

cc @jeffschwMSFT

**Comment by jkotas** [comment]
> @xoofx We are not exposing the raw LoadLibrary or dlopen since it is so simple to just do a P/Invoke yourself. 

It is not simple. It is actually close to impossible to do in portable way on Unix.

**Comment by xoofx** [comment]
> @jkotas I have seen examples - the ones provided are also basically the same. I don't there are developers unloading native DLLs, more of my statement is around 'why?' - what scenario is being unblocked here, other than 'I want this'. The hard numbers I am looking for are performance, memory, some real numbers that move the needle on why this is important.

The scenario is quite common. A load library is locking the library on the disk. You might use a function temporarily but you might want to unlock the dll once you are done with it. If you have recompilation on the fly for some projects and automatic reloading of code, you want to be able to unlock the library. A LoadLibrary is allocating memory, why we would not be able to free it if our application has a requirement for that? How can you judge that this is not important?

> @xoofx We are not exposing the raw LoadLibrary or dlopen since it is so simple to just do a P/Invoke yourself. 

Yes but this is laborious to do, typical example is in the above [code](https://github.com/esteve/ros2_dotnet/blob/e0ea001d2a63a6d23d440aed26f9223dc5459d2a/rcldotnet_common/DllLoadUtils.cs#L123) posted. I have written similar code and it's just stupid to have no cross platform way to have this builtin into the runtime. If this functions were abstracted by as an internal call, we would not have to care about these details. So yeah, sure, we can do it, but why not doing it correctly in coreclr in the first place? We have a bunch of similar low-level functions in `Marshal`, why can't we have basic functions like this?

Again, I'm not against the high level purpose of NativeLibrary, but please, it will cost you almost nothing to decouple it, and it will open more advanced scenarios.

**Comment by mellinoe** [comment]
Just chiming in to confirm that portably invoking dlopen is extremely difficult / impossible. I think exposing dlopen / dlsym / LoadLibrary / GetProcAddress, etc. would be the minimum feature with the highest return on investment. I don’t have a strong opinion on dlclose and friends.

**Comment by xoofx** [comment]
Just to give more context about why I'm looking for a simple low-level API (i.e I'm not here just for a fancy request because it's fun to have it)

At Unity, we have been developing the "burst" compiler that translates .NET assembly code to efficient native code using LLVM. Today we are using LLVM JIT in the editor, so everytime a user is pressing CTRL-S on a cs file in its solution, we need to recompile all the jobs that are being executed in the editor (in Unity, we have a full domain reload, we then discard what was previously loaded). But we are likely gonna have to move to a dll approach for loading these functions, because the LLVM JIT does not provide a good debugging experience (none on Windows). So we have a strong requirement to unload a library. Today, we have our own platform abstraction of LoadLibrary (that would be similar to what I'm asking here), but I would much prefer that we use standard .NET API for this.

We are also not looking for the `DllImportSearchPath` high level feature, because we have a custom layout of our folders and we will need to provide a custom resolution for our shared libraries.

**Comment by swaroop-sridhar** [comment]
Thanks for the feedback. I have updated the text of the proposal to address the feedback above: 
* Clarify that `Load(LibraryName)` is the simple wrapper around `dlOpen()`/`LoadLibrary()`
* Add abstractions for `Free()` and `GetProcAddress()`

**Comment by xoofx** [comment]
Thanks @terrajobst for the update, it looks much better 👍 

I have a few questions:

For `LoadLibrary(string libraryPath)`, I suppose that it will be a simple wrapper around LoadLibrary/dlopen. So `kernel32.dll` would work, as well as a full path right?

But if we pass only `LoadLibrary("kernel32")` it would not work, correct?

Is the version `LoadLibrary(string libraryName, Assembly assembly,                                          DllImportSearchPath? searchPath);` supposed to support this case? (trying to post fix libraryName with platform dependent common for the current platform?).
- I expect this version to use the simpler `LoadLibrary` (or `TryLoadLibrary(libraryPath)`), correct?
- What's the purpose of the assembly passed?

I'm wondering if it would be valuable to expose a method that returns common library extensions for the current platform (or just a string might be enough, but on some platforms, like macos, I think that `.so` and `.dylib`  are possible):

```c#
        public static string[] GetLibraryExtensions();
...

**Comment by xoofx** [comment]
@jkotas Thanks for all the answers.

> LoadLibrary("kernel32") works fine in C/C++ on Windows, so this would work too. This API is a thin wrapper around LoadLibrary/dlopen. It is not going to try to artifically supress any probing logic that the OS API provides.

Yeah, sorry, on Windows it does, but I'm not sure that dlopen does this actually? 
Otherwise, that would be `LoadLibrary(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)` that could handle this.

In the case we would like to have custom dll search path, but still we would like to get the benefit of the auto prefix/postfix and search probe path, wouldn't make sense to pass a list of additional probing paths directly to LoadLibrary instead of an Assembly? (Do we use anything else the `assembly.Location`? behind the door?) Something like:

```c#
        public static IntPtr LoadLibrary(string libraryName,
                                         string[] additionalProbingPaths,
                                         DllImportSearchPath? searchPath);
```

**Comment by jkotas** [comment]
> Yeah, sorry, on Windows it does, but I'm not sure that dlopen does this actually?

Right.

> custom dll search path

We know that there are scenarios that may want to supply custom dll search paths of various forms. We are leaving designing APIs for those for the next iteration.

> Do we use anything else the `assembly.Location`

Yes.

**Comment by eerhardt** [comment]
1. In general, why have both `Get` and `TryGet` overloads? Why can't `IntPtr == 0` signify that it wasn't found? And if it can't, then why not just have the `TryGet` methods?

2. Don't you need to pass in the handle to the library in `GetLibraryExport`? 
```C#
        public static IntPtr GetLibraryExport(string name);
        public static bool TryGetLibraryExport(string name, out IntPtr address);
```

**Comment by xoofx** [comment]
> In general, why have both Get and TryGet overloads? Why can't IntPtr == 0 signify that it wasn't found? And if it can't, then why not just have the TryGet methods?

It may actually be practical to have `LoadLibrary` throwing an exception, as on dlopen platforms, we can get the error message via `dlerror()` right after a failing dlopen (same for dlsym). So, the approach with TryLoad/Load makes sense here. Though, I don't know if dlerror gives relevant details in practice...

> Don't you need to pass in the handle to the library in GetLibraryExport?

yep an oversight likely 😉

**Comment by AaronRobinsonMSFT** [comment]
>Don't you need to pass in the handle to the library in GetLibraryExport?

Updated.

**Comment by swaroop-sridhar** [comment]
There is one more API pending to be discussed: `RegisterDllImportResolver`

**Comment by tannergooding** [comment]
Right, but that shouldn't block the already approved subset, correct?

**Comment by swaroop-sridhar** [comment]
Yes, correct.

**Comment by terrajobst** [comment]
This is the shape we agreed on:

* We centrlized the APIs on a new type `NativeLibrary` as opposed to `Marshal` because it helps the developer using these APIs as it groups them better. `Marshal` already has a ton of APIs.
* We renamed `RegisterDllImportResolver` to `SetDllImportResolver`. Register implies having an unregister and having a chain. `Set` makes it clear there is a single value and `null` will clear it.
* We replaced the func of `SetDllImportResolver` with a named delegate so that we can name the parameters.
* Open item: we need to clarify whether the `DllImportSearchPath?` argument to `DllImportResolver` is different from any other existing enum value. If not, we should make it non-nullable.


```C#
namespace System.Runtime.InteropServices
{
    public delegate IntPtr DllImportResolver(string libraryName,
                                             Assembly assembly,
                                             DllImportSearchPath? searchPath);

    public static partial class NativeLibrary
    {
        // Typical or for dependencies which must be there or failure should happen.

        public static IntPtr Load(string libraryPath);
        public static IntPtr Load(string libraryName,
                                  Assembly assembly,
                                  DllImportSearchPath? searchPath);

        // For fast probing scenarios:

        public static bool TryLoad(string libraryPath,
                                   out IntPtr handle);
        public static bool TryLoad(string libraryName,
                                   Assembly assembly,
                                   DllImportSearchPath? searchPath,
                                   out IntPtr handle);

        public static void Free(IntPtr handle);

        public static IntPtr GetExport(IntPtr handle, string name);
        public static bool TryGetExport(IntPtr handle, string name, out IntPtr address);

        public static bool SetDllImportResolver(Assembly assembly, DllImportResolver resolver);
    }
}
```

**Comment by swaroop-sridhar** [comment]
The [`LoadLibrary`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.defaultdllimportsearchpathsattribute?view=netframework-4.7.2) spec says: 

"The common language runtime handles the call to the `LoadLibraryEx` function according to the following algorithm:

1) If the attribute is applied to an individual platform invoke, use the values specified by that instance of the attribute.

2) Otherwise, if the attribute is applied to the assembly that contains the platform invoke, use the values specified by that instance of the attribute.

3) Otherwise, search the assembly directory and then call the `LoadLibraryEx` function with the `LOAD_WITH_ALTERED_SEARCH_PATH` flag."

Therefore, the difference between default behavior (when no attributes are found) and legacy behavior (when the attribute is present) is that:
1) Legacy Behavior: Search **application** directory then call `LoadLibraryEx` with `LOAD_WITH_ALTERED_SEARCH_PATH` flag.
2) Default behavior: Search **assembly** directory then call `LOAD_WITH_ALTERED_SEARCH_PATH` flag.

So, I think that we should keep the `DllImportResolver` parameter nullable on `SetDllImportResolver` API.

**Comment by swaroop-sridhar** [comment]
Yes DefaultDllImportSearchPath is currently Windows-specific, except for the SearchAssemblyDirectory value. I filed https://github.com/dotnet/coreclr/issues/21584 to track this.

**Comment by jkotas** [comment]
> `bool SetDllImportResolver`

What do we expect the callers to do when this returns false?

**Comment by weltkante** [comment]
> We renamed RegisterDllImportResolver to SetDllImportResolver. Register implies having an unregister and having a chain. Set makes it clear there is a single value and null will clear it.

Having consumers to design around a single global resource sounds to me like a recipe for frustration. How can we prevent 3rd party libraries from calling it? If only a single delegate can be registered it should be reserved to the main application, otherwise libraries are gonna try and take ownership of the resolver for themselves.

**Comment by swaroop-sridhar** [comment]
>> What do we expect the callers to do when this returns false?
There are a few possibilities:
1) Handling depends on the app -- likely indicates a bug in the app? 
2) Unset the resolver by passing `nullptr` handle. Then they can pass a new resolver. This makes the overwrite explicit, but may be confusing to use. 

@jkotas, Since the API is now called `SetDllImportResolver` I think it is more intuitive to always overwrite the resolver per-assembly (or unset by setting `nullptr`). We could add `GetDllImportResolver`  or `HasDllImportResolver`  properties if deemed necessary.

**Comment by jkotas** [comment]
> Handling depends on the app -- likely indicates a bug in the app?

Right. We just throw exceptions for patterns that are bugs in the app. We do not return `bool`.

> Unset the resolver by passing nullptr handle.

Why would anybody want to do that? It would be just a factory for race conditions and other hard to diagnose problems.

**Comment by jeffschwMSFT** [comment]
Closed with #34686
